File: memorylayer-sdk-python/src/memorylayer/mcp_servers.py
"""MCP servers namespace for MemoryLayer.ai Python SDK.

Provides CRUD, resolve, and push/pull helpers for MCP server records.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING, Any, Literal

from pydantic import BaseModel, ConfigDict, Field

if TYPE_CHECKING:
    from .client import MemoryLayerClient
    from .models import AuthorityContext
    from .sync_client import SyncMemoryLayerClient


class McpServerModel(BaseModel):
    """Pydantic model mirroring the server McpServer."""

    model_config = ConfigDict(extra="allow")

    id: str
    workspace_id: str
    tenant_id: str = "_default"
    user_id: str | None = None
    name: str
    description: str | None = None
    transport: str
    command: str | None = None
    args: list[str] = Field(default_factory=list)
    env: dict[str, str] = Field(default_factory=dict)
    url: str | None = None
    headers: dict[str, str] = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)
    source_mode: str = "server"
    manifest_hash: str = ""
    enabled: bool = True
    created_at: str | None = None
    updated_at: str | None = None


class McpServersAPI:
    """MCP servers namespace — access via client.mcp_servers.<method>."""

    def __init__(self, client: MemoryLayerClient) -> None:
        self._client = client

    def _ws(self, workspace_id: str | None) -> str | None:
        return workspace_id or self._client.workspace_id

    # ------------------------------------------------------------------
    # CRUD
    # ------------------------------------------------------------------

    async def list(
        self,
        transport: str | None = None,
        enabled: bool | None = None,
        name: str | None = None,
        user_id: str | None = None,
        limit: int = 100,
        offset: int = 0,
        workspace_id: str | None = None,
        authority: AuthorityContext | None = None,
    ) -> list[McpServerModel]:
        """List MCP servers with optional filters."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        ws_id = self._ws(workspace_id)
        if ws_id:
            params["workspace_id"] = ws_id
        if transport is not None:
            params["transport"] = transport
        if enabled is not None:
            params["enabled"] = str(enabled).lower()
        if name is not None:
            params["name"] = name
        if user_id is not None:
            params["user_id"] = user_id

        data = await self._client._request("GET", "/mcp-servers", params=params, authority=authority)
        return [McpServerModel(**s) for s in data.get("mcp_servers", [])]

    async def get(self, server_id: str, authority: AuthorityContext | None = None) -> McpServerModel:
        """Get an MCP server by ID."""
        data = await self._client._request("GET", f"/mcp-servers/{server_id}", authority=authority)
        return McpServerModel(**data["mcp_server"])

    async def get_by_name(
        self,
        name: str,
        user_id: str | None = None,
        workspace_id: str | None = None,
        authority: AuthorityContext | None = None,
    ) -> McpServerModel | None:
        """Get an MCP server by name; returns None if not found."""
        params: dict[str, Any] = {"name": name}
        ws_id = self._ws(workspace_id)
        if ws_id:
            params["workspace_id"] = ws_id
        if user_id is not None:
            params["user_id"] = user_id

        try:
            data = await self._client._request("GET", "/mcp-servers/by-name", params=params, authority=authority)
            return McpServerModel(**data["mcp_server"])
        except Exception:
            return None

    async def create(
        self,
        name: str,
        transport: Literal["stdio", "http", "sse", "streamable-http"],
        command: str | None = None,
        args: list[str] | None = None,
        env: dict[str, str] | None = None,
        url: str | None = None,
        headers: dict[str, str] | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
        source_mode: str = "server",
        enabled: bool = True,
        workspace_id: str | None = None,
        user_id: str | None = None,
        authority: AuthorityContext | None = None,
    ) -> McpServerModel:
        """Create a new MCP server record."""
        payload: dict[str, Any] = {
            "name": name,
            "transport": transport,
            "source_mode": source_mode,
            "enabled": enabled,
        }
        ws_id = self._ws(workspace_id)
        if ws_id:
            payload["workspace_id"] = ws_id
        if user_id is not None:
            payload["user_id"] = user_id
        if description is not None:
            payload["description"] = description
        if command is not None:
            payload["command"] = command
        if args is not None:
            payload["args"] = args
        if env is not None:
            payload["env"] = env
        if url is not None:
            payload["url"] = url
        if headers is not None:
            payload["headers"] = headers
        if metadata is not None:
            payload["metadata"] = metadata

        data = await self._client._request("POST", "/mcp-servers", json=payload, authority=authority)
        return McpServerModel(**data["mcp_server"])

    async def update(
        self,
        server_id: str,
        description: str | None = None,
        command: str | None = None,
        args: list[str] | None = None,
        env: dict[str, str] | None = None,
        url: str | None = None,
        headers: dict[str, str] | None = None,
        metadata: dict[str, Any] | None = None,
        source_mode: str | None = None,
        enabled: bool | None = None,
        authority: AuthorityContext | None = None,
    ) -> McpServerModel:
        """Partially update an MCP server record."""
        payload: dict[str, Any] = {}
        if description is not None:
            payload["description"] = description
        if command is not None:
            payload["command"] = command
        if args is not None:
            payload["args"] = args
        if env is not None:
            payload["env"] = env
        if url is not None:
            payload["url"] = url
        if headers is not None:
            payload["headers"] = headers
        if metadata is not None:
            payload["metadata"] = metadata
        if source_mode is not None:
            payload["source_mode"] = source_mode
        if enabled is not None:
            payload["enabled"] = enabled

        data = await self._client._request("PUT", f"/mcp-servers/{server_id}", json=payload, authority=authority)
        return McpServerModel(**data["mcp_server"])

    async def delete(self, server_id: str, authority: AuthorityContext | None = None) -> None:
        """Delete an MCP server record."""
        await self._client._request("DELETE", f"/mcp-servers/{server_id}", authority=authority)

    async def resolve(
        self,
        name: str,
        workspace_id: str | None = None,
        authority: AuthorityContext | None = None,
    ) -> McpServerModel | None:
        """Resolve an MCP server by name using 4-tier precedence; returns None if not found."""
        payload: dict[str, Any] = {"name": name}
        ws_id = self._ws(workspace_id)
        if ws_id:
            payload["workspace_id"] = ws_id

        data = await self._client._request("POST", "/mcp-servers/resolve", json=payload, authority=authority)
        server = data.get("mcp_server")
        if server is not None:
            return McpServerModel(**server)
        return None

    # ------------------------------------------------------------------
    # .mcp.json helpers
    # ------------------------------------------------------------------

    async def push_json(
        self,
        json_path: Path,
        scope: str = "workspace",
        source_mode: str = "server",
        workspace_id: str | None = None,
        user_id: str | None = None,
        authority: AuthorityContext | None = None,
    ) -> list[McpServerModel]:
        """Push a .mcp.json file to MemoryLayer, creating/updating one record per server entry."""
        json_path = Path(json_path)
        doc = json.loads(json_path.read_text(encoding="utf-8"))
        mcp_servers = doc.get("mcpServers", {})

        results: list[McpServerModel] = []
        ws_id = self._ws(workspace_id)
        for name, entry in mcp_servers.items():
            transport = entry.get("transport")
            if transport is None:
                transport = "stdio" if entry.get("command") else "http"

            server = await self.create(
                name=name,
                transport=transport,
                command=entry.get("command"),
                args=entry.get("args"),
                env=entry.get("env"),
                url=entry.get("url"),
                headers=entry.get("headers"),
                source_mode=source_mode,
                workspace_id=ws_id,
                user_id=user_id,
                authority=authority,
            )
            results.append(server)
        return results

    async def pull_json(
        self,
        out_path: Path,
        workspace_id: str | None = None,
        user_id: str | None = None,
        transport: str | None = None,
        enabled: bool | None = None,
        authority: AuthorityContext | None = None,
    ) -> Path:
        """Materialize MCP servers to a .mcp.json file at out_path."""
        servers = await self.list(
            transport=transport,
            enabled=enabled,
            user_id=user_id,
            workspace_id=workspace_id,
            authority=authority,
        )

        mcp_servers: dict[str, Any] = {}
        for server in servers:
            entry: dict[str, Any] = {"transport": server.transport}
            if server.transport == "stdio":
                if server.command:
                    entry["command"] = server.command
                if server.args:
                    entry["args"] = server.args
                if server.env:
                    entry["env"] = server.env
            else:
                if server.url:
                    entry["url"] = server.url
                if server.headers:
                    entry["headers"] = server.headers
            mcp_servers[server.name] = entry

        out_path = Path(out_path)
        out_path.write_text(
            json.dumps({"mcpServers": mcp_servers}, indent=2),
            encoding="utf-8",
        )
        return out_path


class SyncMcpServersAPI:
    """Synchronous MCP servers namespace — access via sync_client.mcp_servers.<method>."""

    def __init__(self, client: SyncMemoryLayerClient) -> None:
        self._client = client

    def _ws(self, workspace_id: str | None) -> str | None:
        return workspace_id or self._client.workspace_id

    def list(
        self,
        transport: str | None = None,
        enabled: bool | None = None,
        name: str | None = None,
        user_id: str | None = None,
        limit: int = 100,
        offset: int = 0,
        workspace_id: str | None = None,
    ) -> list[McpServerModel]:
        """List MCP servers with optional filters."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        ws_id = self._ws(workspace_id)
        if ws_id:
            params["workspace_id"] = ws_id
        if transport is not None:
            params["transport"] = transport
        if enabled is not None:
            params["enabled"] = str(enabled).lower()
        if name is not None:
            params["name"] = name
        if user_id is not None:
            params["user_id"] = user_id

        data = self._client._request("GET", "/mcp-servers", params=params)
        return [McpServerModel(**s) for s in data.get("mcp_servers", [])]

    def get(self, server_id: str) -> McpServerModel:
        """Get an MCP server by ID."""
        data = self._client._request("GET", f"/mcp-servers/{server_id}")
        return McpServerModel(**data["mcp_server"])

    def get_by_name(
        self,
        name: str,
        user_id: str | None = None,
        workspace_id: str | None = None,
    ) -> McpServerModel | None:
        """Get an MCP server by name; returns None if not found."""
        params: dict[str, Any] = {"name": name}
        ws_id = self._ws(workspace_id)
        if ws_id:
            params["workspace_id"] = ws_id
        if user_id is not None:
            params["user_id"] = user_id

        try:
            data = self._client._request("GET", "/mcp-servers/by-name", params=params)
            return McpServerModel(**data["mcp_server"])
        except Exception:
            return None

    def create(
        self,
        name: str,
        transport: Literal["stdio", "http", "sse", "streamable-http"],
        command: str | None = None,
        args: list[str] | None = None,
        env: dict[str, str] | None = None,
        url: str | None = None,
        headers: dict[str, str] | None = None,
        description: str | None = None,
        metadata: dict[str, Any] | None = None,
        source_mode: str = "server",
        enabled: bool = True,
        workspace_id: str | None = None,
        user_id: str | None = None,
    ) -> McpServerModel:
        """Create a new MCP server record."""
        payload: dict[str, Any] = {
            "name": name,
            "transport": transport,
            "source_mode": source_mode,
            "enabled": enabled,
        }
        ws_id = self._ws(workspace_id)
        if ws_id:
            payload["workspace_id"] = ws_id
        if user_id is not None:
            payload["user_id"] = user_id
        if description is not None:
            payload["description"] = description
        if command is not None:
            payload["command"] = command
        if args is not None:
            payload["args"] = args
        if env is not None:
            payload["env"] = env
        if url is not None:
            payload["url"] = url
        if headers is not None:
            payload["headers"] = headers
        if metadata is not None:
            payload["metadata"] = metadata

        data = self._client._request("POST", "/mcp-servers", json=payload)
        return McpServerModel(**data["mcp_server"])

    def update(
        self,
        server_id: str,
        description: str | None = None,
        command: str | None = None,
        args: list[str] | None = None,
        env: dict[str, str] | None = None,
        url: str | None = None,
        headers: dict[str, str] | None = None,
        metadata: dict[str, Any] | None = None,
        source_mode: str | None = None,
        enabled: bool | None = None,
    ) -> McpServerModel:
        """Partially update an MCP server record."""
        payload: dict[str, Any] = {}
        if description is not None:
            payload["description"] = description
        if command is not None:
            payload["command"] = command
        if args is not None:
            payload["args"] = args
        if env is not None:
            payload["env"] = env
        if url is not None:
            payload["url"] = url
        if headers is not None:
            payload["headers"] = headers
        if metadata is not None:
            payload["metadata"] = metadata
        if source_mode is not None:
            payload["source_mode"] = source_mode
        if enabled is not None:
            payload["enabled"] = enabled

        data = self._client._request("PUT", f"/mcp-servers/{server_id}", json=payload)
        return McpServerModel(**data["mcp_server"])

    def delete(self, server_id: str) -> None:
        """Delete an MCP server record."""
        self._client._request("DELETE", f"/mcp-servers/{server_id}")

    def resolve(
        self,
        name: str,
        workspace_id: str | None = None,
    ) -> McpServerModel | None:
        """Resolve an MCP server by name using 4-tier precedence; returns None if not found."""
        payload: dict[str, Any] = {"name": name}
        ws_id = self._ws(workspace_id)
        if ws_id:
            payload["workspace_id"] = ws_id

        data = self._client._request("POST", "/mcp-servers/resolve", json=payload)
        server = data.get("mcp_server")
        if server is not None:
            return McpServerModel(**server)
        return None

    def push_json(
        self,
        json_path: Path,
        scope: str = "workspace",
        source_mode: str = "server",
        workspace_id: str | None = None,
        user_id: str | None = None,
    ) -> list[McpServerModel]:
        """Push a .mcp.json file to MemoryLayer, creating/updating one record per server entry."""
        json_path = Path(json_path)
        doc = json.loads(json_path.read_text(encoding="utf-8"))
        mcp_servers = doc.get("mcpServers", {})

        results: list[McpServerModel] = []
        ws_id = self._ws(workspace_id)
        for name, entry in mcp_servers.items():
            transport = entry.get("transport")
            if transport is None:
                transport = "stdio" if entry.get("command") else "http"

            server = self.create(
                name=name,
                transport=transport,
                command=entry.get("command"),
                args=entry.get("args"),
                env=entry.get("env"),
                url=entry.get("url"),
                headers=entry.get("headers"),
                source_mode=source_mode,
                workspace_id=ws_id,
                user_id=user_id,
            )
            results.append(server)
        return results

    def pull_json(
        self,
        out_path: Path,
        workspace_id: str | None = None,
        user_id: str | None = None,
        transport: str | None = None,
        enabled: bool | None = None,
    ) -> Path:
        """Materialize MCP servers to a .mcp.json file at out_path."""
        servers = self.list(
            transport=transport,
            enabled=enabled,
            user_id=user_id,
            workspace_id=workspace_id,
        )

        mcp_servers: dict[str, Any] = {}
        for server in servers:
            entry: dict[str, Any] = {"transport": server.transport}
            if server.transport == "stdio":
                if server.command:
                    entry["command"] = server.command
                if server.args:
                    entry["args"] = server.args
                if server.env:
                    entry["env"] = server.env
            else:
                if server.url:
                    entry["url"] = server.url
                if server.headers:
                    entry["headers"] = server.headers
            mcp_servers[server.name] = entry

        out_path = Path(out_path)
        out_path.write_text(
            json.dumps({"mcpServers": mcp_servers}, indent=2),
            encoding="utf-8",
        )
        return out_path

File: memorylayer-sdk-go/mcp_servers.go
package memorylayer

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"os"
	"strconv"
)

// McpServerModel mirrors the server McpServer record.
type McpServerModel struct {
	ID           string            `json:"id"`
	WorkspaceID  string            `json:"workspace_id"`
	TenantID     string            `json:"tenant_id"`
	UserID       *string           `json:"user_id,omitempty"`
	Name         string            `json:"name"`
	Description  *string           `json:"description,omitempty"`
	Transport    string            `json:"transport"`
	Command      *string           `json:"command,omitempty"`
	Args         []string          `json:"args"`
	Env          map[string]string `json:"env"`
	URL          *string           `json:"url,omitempty"`
	Headers      map[string]string `json:"headers"`
	Metadata     map[string]any    `json:"metadata"`
	SourceMode   string            `json:"source_mode"`
	ManifestHash string            `json:"manifest_hash"`
	Enabled      bool              `json:"enabled"`
	CreatedAt    *string           `json:"created_at,omitempty"`
	UpdatedAt    *string           `json:"updated_at,omitempty"`
}

// McpServersAPI is the /v1/mcp-servers namespace, reached via client.McpServers.
type McpServersAPI struct {
	client *Client
}

func (m *McpServersAPI) ws(workspaceID string) string {
	return m.client.resolveWorkspace(workspaceID)
}

// ListMcpServersOptions configures [McpServersAPI.List].
type ListMcpServersOptions struct {
	Transport   string
	Enabled     *bool
	Name        string
	UserID      string
	Limit       int // default 100 when <= 0
	Offset      int
	WorkspaceID string
	Authority   *AuthorityContext
}

// List lists MCP servers with optional filters.
func (m *McpServersAPI) List(ctx context.Context, opts ListMcpServersOptions) ([]McpServerModel, error) {
	limit := opts.Limit
	if limit <= 0 {
		limit = 100
	}
	q := url.Values{"limit": {strconv.Itoa(limit)}, "offset": {strconv.Itoa(opts.Offset)}}
	if ws := m.ws(opts.WorkspaceID); ws != "" {
		q.Set("workspace_id", ws)
	}
	if opts.Transport != "" {
		q.Set("transport", opts.Transport)
	}
	if opts.Enabled != nil {
		q.Set("enabled", boolStr(*opts.Enabled))
	}
	if opts.Name != "" {
		q.Set("name", opts.Name)
	}
	if opts.UserID != "" {
		q.Set("user_id", opts.UserID)
	}
	var out struct {
		McpServers []McpServerModel `json:"mcp_servers"`
	}
	if err := m.client.doJSON(ctx, requestSpec{
		method:    http.MethodGet,
		path:      "/mcp-servers",
		query:     q,
		authority: opts.Authority,
	}, &out); err != nil {
		return nil, err
	}
	return out.McpServers, nil
}

// Get fetches an MCP server by ID.
func (m *McpServersAPI) Get(ctx context.Context, serverID string, authority *AuthorityContext) (*McpServerModel, error) {
	var out struct {
		McpServer McpServerModel `json:"mcp_server"`
	}
	if err := m.client.doJSON(ctx, requestSpec{
		method:    http.MethodGet,
		path:      "/mcp-servers/" + serverID,
		authority: authority,
	}, &out); err != nil {
		return nil, err
	}
	return &out.McpServer, nil
}

// CreateMcpServerOptions configures [McpServersAPI.Create].
type CreateMcpServerOptions struct {
	Command     string
	Args        []string
	Env         map[string]string
	URL         string
	Headers     map[string]string
	Description string
	Metadata    map[string]any
	SourceMode  string // default "server"
	Enabled     *bool  // default true
	WorkspaceID string
	UserID      string
	Authority   *AuthorityContext
}

// Create creates a new MCP server record. transport is one of "stdio", "http",
// "sse", or "streamable-http".
func (m *McpServersAPI) Create(ctx context.Context, name, transport string, opts CreateMcpServerOptions) (*McpServerModel, error) {
	enabled := true
	if opts.Enabled != nil {
		enabled = *opts.Enabled
	}
	payload := map[string]any{
		"name":        name,
		"transport":   transport,
		"source_mode": orDefault(opts.SourceMode, "server"),
		"enabled":     enabled,
	}
	if ws := m.ws(opts.WorkspaceID); ws != "" {
		payload["workspace_id"] = ws
	}
	if opts.UserID != "" {
		payload["user_id"] = opts.UserID
	}
	if opts.Description != "" {
		payload["description"] = opts.Description
	}
	if opts.Command != "" {
		payload["command"] = opts.Command
	}
	if opts.Args != nil {
		payload["args"] = opts.Args
	}
	if opts.Env != nil {
		payload["env"] = opts.Env
	}
	if opts.URL != "" {
		payload["url"] = opts.URL
	}
	if opts.Headers != nil {
		payload["headers"] = opts.Headers
	}
	if opts.Metadata != nil {
		payload["metadata"] = opts.Metadata
	}
	var out struct {
		McpServer McpServerModel `json:"mcp_server"`
	}
	if err := m.client.requestJSON(ctx, http.MethodPost, "/mcp-servers", payload, requestSpec{authority: opts.Authority}, &out); err != nil {
		return nil, err
	}
	return &out.McpServer, nil
}

// UpdateMcpServerOptions configures [McpServersAPI.Update]. Nil/zero fields are
// omitted (unchanged).
type UpdateMcpServerOptions struct {
	Description *string
	Command     *string
	Args        []string
	Env         map[string]string
	URL         *string
	Headers     map[string]string
	Metadata    map[string]any
	SourceMode  *string
	Enabled     *bool
	Authority   *AuthorityContext
}

// Update partially updates an MCP server record.
func (m *McpServersAPI) Update(ctx context.Context, serverID string, opts UpdateMcpServerOptions) (*McpServerModel, error) {
	payload := map[string]any{}
	if opts.Description != nil {
		payload["description"] = *opts.Description
	}
	if opts.Command != nil {
		payload["command"] = *opts.Command
	}
	if opts.Args != nil {
		payload["args"] = opts.Args
	}
	if opts.Env != nil {
		payload["env"] = opts.Env
	}
	if opts.URL != nil {
		payload["url"] = *opts.URL
	}
	if opts.Headers != nil {
		payload["headers"] = opts.Headers
	}
	if opts.Metadata != nil {
		payload["metadata"] = opts.Metadata
	}
	if opts.SourceMode != nil {
		payload["source_mode"] = *opts.SourceMode
	}
	if opts.Enabled != nil {
		payload["enabled"] = *opts.Enabled
	}
	var out struct {
		McpServer McpServerModel `json:"mcp_server"`
	}
	if err := m.client.requestJSON(ctx, http.MethodPut, "/mcp-servers/"+serverID, payload, requestSpec{authority: opts.Authority}, &out); err != nil {
		return nil, err
	}
	return &out.McpServer, nil
}

// Delete deletes an MCP server record.
func (m *McpServersAPI) Delete(ctx context.Context, serverID string, authority *AuthorityContext) error {
	return m.client.doJSON(ctx, requestSpec{
		method:    http.MethodDelete,
		path:      "/mcp-servers/" + serverID,
		authority: authority,
	}, nil)
}

// Resolve resolves an MCP server by name using 4-tier precedence; returns
// (nil, nil) when not found.
func (m *McpServersAPI) Resolve(ctx context.Context, name, workspaceID string, authority *AuthorityContext) (*McpServerModel, error) {
	payload := map[string]any{"name": name}
	if ws := m.ws(workspaceID); ws != "" {
		payload["workspace_id"] = ws
	}
	var out struct {
		McpServer *McpServerModel `json:"mcp_server"`
	}
	if err := m.client.requestJSON(ctx, http.MethodPost, "/mcp-servers/resolve", payload, requestSpec{authority: authority}, &out); err != nil {
		return nil, err
	}
	return out.McpServer, nil
}

// ------------------------------------------------------------------
// .mcp.json helpers
// ------------------------------------------------------------------

type mcpJSONEntry struct {
	Transport string            `json:"transport,omitempty"`
	Command   string            `json:"command,omitempty"`
	Args      []string          `json:"args,omitempty"`
	Env       map[string]string `json:"env,omitempty"`
	URL       string            `json:"url,omitempty"`
	Headers   map[string]string `json:"headers,omitempty"`
}

type mcpJSONDoc struct {
	McpServers map[string]mcpJSONEntry `json:"mcpServers"`
}

// PushJSON pushes a .mcp.json file, creating one record per server entry.
func (m *McpServersAPI) PushJSON(ctx context.Context, jsonPath, scope, sourceMode, workspaceID, userID string, authority *AuthorityContext) ([]McpServerModel, error) {
	data, err := os.ReadFile(jsonPath)
	if err != nil {
		return nil, err
	}
	var doc mcpJSONDoc
	if err := json.Unmarshal(data, &doc); err != nil {
		return nil, err
	}
	var results []McpServerModel
	for name, entry := range doc.McpServers {
		transport := entry.Transport
		if transport == "" {
			if entry.Command != "" {
				transport = "stdio"
			} else {
				transport = "http"
			}
		}
		server, err := m.Create(ctx, name, transport, CreateMcpServerOptions{
			Command:     entry.Command,
			Args:        entry.Args,
			Env:         entry.Env,
			URL:         entry.URL,
			Headers:     entry.Headers,
			SourceMode:  sourceMode,
			WorkspaceID: workspaceID,
			UserID:      userID,
			Authority:   authority,
		})
		if err != nil {
			return nil, err
		}
		results = append(results, *server)
	}
	return results, nil
}

// PullJSON materializes MCP servers into a .mcp.json file at outPath.
func (m *McpServersAPI) PullJSON(ctx context.Context, outPath, workspaceID, userID, transport string, enabled *bool, authority *AuthorityContext) error {
	servers, err := m.List(ctx, ListMcpServersOptions{
		Transport:   transport,
		Enabled:     enabled,
		UserID:      userID,
		WorkspaceID: workspaceID,
		Authority:   authority,
	})
	if err != nil {
		return err
	}
	doc := mcpJSONDoc{McpServers: map[string]mcpJSONEntry{}}
	for _, s := range servers {
		entry := mcpJSONEntry{Transport: s.Transport}
		if s.Transport == "stdio" {
			if s.Command != nil {
				entry.Command = *s.Command
			}
			entry.Args = s.Args
			entry.Env = s.Env
		} else {
			if s.URL != nil {
				entry.URL = *s.URL
			}
			entry.Headers = s.Headers
		}
		doc.McpServers[s.Name] = entry
	}
	body, err := json.MarshalIndent(doc, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(outPath, body, 0o644)
}
